pkg/auth/oidc_connect.go

```diff
@@ -0,0 +1,294 @@
+
+	state := store.NewState()
+	state.GenerateOauthState()
+	state.RequestPath = u.Host + u.Path
+
+	resp.Response.Header.Add(oauthTokenName, "")
+
+	authCodeURL := h.oauth2Config().AuthCodeURL(state.OAuthState)
+	byteState := store.ConvertToByte(state)
+
+	h.Cache.Set(state.OAuthState, byteState)
+	resp.Response.StatusCode = http.StatusFound
+	resp.Response.Header.Add("Location", authCodeURL)
+}
+
+func (h *OIDCConnect) callbackHandler(ctx context.Context, req *Request, resp *Response, u *url.URL) {
+
+	// 1. Get all variable needed.
+	oauthState := u.Query().Get("state")
+	code := u.Query().Get("code")
+
+	// Check state and code validity.
+	if code == "" || oauthState == "" {
+		resp.Response.StatusCode = http.StatusBadRequest
+		resp.Allow = false
+		return
+	}
+
+	stateByte, err := h.Cache.Get(oauthState)
+	if err != nil {
+		h.Log.Info("Invalid state, expected " + oauthState + " not found ")
+		resp.Response.StatusCode = http.StatusBadRequest
+		resp.Allow = false
+		return
+	}
+
+	// Retrieve token.
+	context := oidc.ClientContext(ctx, h.HTTPClient)
+	token, err := h.oauth2Config().Exchange(context, code)
+	if err != nil {
+		h.Log.Info(err.Error())
+	}
+
+	rawIDToken, ok := token.Extra("id_token").(string)
+
+	if !ok {
+		resp.Response.StatusCode = http.StatusInternalServerError
+		resp.Allow = false
+		return
+	}
+
+	//Store token.
+	state := store.ConvertToType(stateByte)
+	state.IDToken = rawIDToken
+	state.AccessToken = token.AccessToken
+	state.RefreshToken = token.RefreshToken
+	stateByte = store.ConvertToByte(state)
+	h.Cache.Set(state.OAuthState, stateByte)
+
+	// Set  redirection.
+	resp.Response.StatusCode = http.StatusTemporaryRedirect
+	resp.Response.Header.Add(stateQueryParamName, state.OAuthState)
+
+	// TODO .. may need to rebuild this, in case user could have other query parameters
+	resp.Response.Header.Add("Location", fmt.Sprintf("http://%s?%s=%s", state.RequestPath, stateQueryParamName, state.OAuthState))
+}
+
+func (h *OIDCConnect) isStateValid(ctx context.Context, state *store.OIDCState, provider *oidc.Provider) bool {
+
+	if state == nil {
+		return false
+	}
+
+	verifier := provider.Verifier(&oidc.Config{ClientID: h.OidcConfig.ClientID, SkipIssuerCheck: true})
+
+	// Verify token and signature.
+	idToken, err := verifier.Verify(ctx, state.IDToken)
+	if err != nil {
+		h.Log.Info(fmt.Sprintf("failed to verify ID token: %v", err))
+		return false
+	}
+
+	// Try to claim.
+	var claims json.RawMessage
+	if err := idToken.Claims(&claims); err != nil {
+		h.Log.Info(fmt.Sprintf("error decoding ID token claims: %v", err))
+		return false
+	}
+
+	return true
+}
+
+func (h *OIDCConnect) getStateFromCookie(ctx context.Context, req *Request) (*store.OIDCState, error) {
+
+	// Do you have cookies stored ?
+	cookieVal := req.Request.Header.Get("cookie")
+	var state *store.OIDCState
+	// Check through and get the right cookies
+	if len(cookieVal) > 0 {
```

Comment:
Could use [Request.Cookie](https://golang.org/pkg/net/http/#Request.Cookie) here?

Reply:
always give me empty if I call that method, thus I did it the crude way